sphinx/ext/autodoc/__init__.py

```diff
@@ -1062,56 +1062,34 @@ def document_members(self, all_members: bool = False) -> None:
 
     def add_directive_header(self, sig: str) -> None:
         sourcename = self.get_sourcename()
+        if inspect.is_singledispatch_function(self.object):
+            # Set the flag so add_line will add the signatures
+            # of the other implementations
+            self._singledispatch_flag = True
+
         super().add_directive_header(sig)
 
         if inspect.iscoroutinefunction(self.object):
             self.add_line('   :async:', sourcename)
 
-
-class SingledispatchFunctionDocumenter(FunctionDocumenter):
-    """
-    Specialized Documenter subclass for singledispatch'ed functions.
-    """
-    objtype = 'singledispatch_function'
-    directivetype = 'function'
-    member_order = 30
-
-    # before FunctionDocumenter
-    priority = FunctionDocumenter.priority + 1
-
-    @classmethod
-    def can_document_member(cls, member: Any, membername: str, isattr: bool, parent: Any
-                            ) -> bool:
-        return (super().can_document_member(member, membername, isattr, parent) and
-                inspect.is_singledispatch_function(member))
-
-    def add_directive_header(self, sig: str) -> None:
-        sourcename = self.get_sourcename()
-
-        # intercept generated directive headers
-        # TODO: It is very hacky to use mock to intercept header generation
-        with patch.object(self, 'add_line') as add_line:
-            super().add_directive_header(sig)
-
-        # output first line of header
-        self.add_line(*add_line.call_args_list[0][0])
+    def add_line(self, line: str, source: str, *lineno: int) -> None:
```

Comment:
-1 for such hack. I agree with my `patch.object()` hack is very ugly. But this is also. I can't accept that the "add_line" method generates content internally.

Reply:
Ok, how about this, I added a `format_extra_signature` to the `Documenter` and overrode that instead, is that any better?